kmsprint: Decode bitfields for flags and type of hdmi mode

popcornmix opened the pull request:
The hex numbers for hdmi mode type and flags are hard to read.
Decode them into something more readable (but still try to keep output concise).

If any field fails to be decoded, then revert to just displaying the hex code
(if this occurs, perhaps due to kernel side extensions we should update the code here).

popcornmix commented:
I'm aware of the duplication of `ARRAY_SIZE` and `join`, but it didn't seem easy to make kms++ depend on kms++util (when there's already a backwards dependency).

I'm happy to change that if you have a preferred way of solving this.

tomba commented on kms++/src/card.cpp:
```diff
@@ -216,6 +216,10 @@ void Card::setup()
+	if (getenv("KMSXX_ENABLE_CLIENT_CAP_STEREO_3D") != 0) {
+		drmSetClientCap(m_fd, DRM_CLIENT_CAP_STEREO_3D, 1);
+	}
+
```
This is not how the other ones work. The idea with KMSXX_DISABLE_UNIVERSAL_PLANES and KMSXX_DISABLE_ATOMIC is that by default the features are enabled. They are disabled only if a) the feature is not available (drmSetClientCap fails), or b) the env variable is set, to explicitly disable the feature.

I'm not familiar with STEREO_3D. What does it mean if it's enabled? What changes? My guess is that we should just enable it by default.

popcornmix replied:
By default only mono modes are reported from kernel. The STEREO_3D capability also reports the 3d variants.
E.g. here is output from a 3D capable TV on Pi without the cap:
http://sprunge.us/NSRzeY
And here with the 3D cap:
http://sprunge.us/qP1frS
You get an extra 14 modes reported here, 
It could be enabled by default, but it creates some extra noise (assuming most users don't care about 3D).

tomba replied:
In that case perhaps it's better to enable 3d by default in kms++, but in kmsprint only show the 3d modes if a cmdline parameter ("-3d"?) is given.

That said, at least in your example, it doesn't look that bad. So I think we might as well have 3d always enabled and kmsprint shows them all. If at some point the extra noise feels too much, we can add the -3d parameter to kmsprint.

tomba commented on kms++/src/videomode.cpp:
```diff
@@ -110,18 +110,111 @@ static char sync_to_char(SyncPolarity pol)
+	NULL, "4:3", "16:9", "64:27", "256:135",
+};
+
+const string mode_type_str(uint32_t inval) {
```
No point in returning "const string". I think you mean "static string".

tomba replied:
"inval" means "invalid" =). Maybe "val" as the parameter name, and "v" as the local variable.

tomba commented on kms++/src/videomode.cpp:
```diff
@@ -110,18 +110,111 @@ static char sync_to_char(SyncPolarity pol)
+		}
+	}
+	if (val != 0)
+		return fmt::format("0x{:x}", inval);
```
This is one way to handle this. But how about adding the val as a string to the vector? Then we'd get something like
"P|D|0x70" where 0x70 are the unhandled flags.

tomba commented on kms++/src/videomode.cpp:
```diff
@@ -110,18 +110,111 @@ static char sync_to_char(SyncPolarity pol)
+	for (size_t i = 0; i < ARRAY_SIZE(mode_type_names); i++) {
+		if (val & (1 << i)) {
+			v.push_back(mode_type_names[i]);
+			val &= ~ (1 << i);
```
Extra space above after ~

tomba commented on kms++/src/videomode.cpp:
```diff
@@ -110,18 +110,111 @@ static char sync_to_char(SyncPolarity pol)
+
+static const char *mode_aspect_names[] = {
+	NULL, "4:3", "16:9", "64:27", "256:135",
+};
```
I'm not very enthusiastic about const char * arrays like these. Would you be ok with similar std::map approach as done in, e.g. connector.cpp? I think it would make the code simpler, especially the mode flags handling, as you only need one mode-map, instead of three separate arrays like you have here.

popcornmix commented:
I've updated. I've dropped the 3D cap for now.
I've switched to a map for the bitfield to string lookups.

> as you only need one mode-map, instead of three separate arrays like you have here

I didn't quite follow this. The aspect and 3d flags can't be handled in the same loop as the single bits flags as they also need different masks to be applied. If you explain exactly how you want this I can try updating this.

tomba commented:
Ah, sorry, my bad. There's just a single 3D value, it wasn't a set of bit flags. So never mind that comment.

tomba commented on kms++/src/videomode.cpp:
```diff
@@ -110,18 +110,123 @@ static char sync_to_char(SyncPolarity pol)
+					s.push_back(x.second);
+				val &= ~x.first;
+			}
+	}
```
This is double indented.

The code elsewhere in kms++ uses "const type&" style, not "type const&".

You can do `for (const auto& [k, v] : mode_type_map)`, which makes iterating maps much nicer (feel free to use other names than k and v).

tomba commented:
I had a few cosmetic comments, but otherwise looks good to me.

popcornmix commented:
Updated to address comments. I also narrowed the type field width to 5 (there are only 3 non-deprecated options, so "P|U|D" is the longest string we are expecting.

tomba commented:
Did you push? I don't see the changes.

popcornmix commented:
Sorry - try now.

tomba commented:
The indentation in the if()s was still wrong, but I fixed that while applying.

popcornmix commented:
Thanks.
